### PR TITLE
Added NLB. Helpers choose to run on labeled Pods

### PR DESCRIPTION
**Pull request by cberkhoff**

### CODE REVIEW

File: eksctl/cluster-config.yaml [hunk 1]
```diff
@@ -6,9 +6,18 @@ metadata:
...
     instanceType: t3.medium
-    desiredCapacity: 2
+    desiredCapacity: 1
+  - name: helper2
+    labels: { releaseName: h2 }
+    instanceType: t3.medium
+    desiredCapacity: 1
+  - name: helper3
+    labels: { releaseName: h3 }
+    instanceType: t3.medium
```
Comment (eriktaubeneck) [inline eksctl/cluster-config.yaml:19]: we aren't utilizing our current boxes completely, but we may want larger boxes than t3.medium. I think we want compute optimized, so maybe `c7i.large` or `c7i.xlarge`?

File: templates/deployment.yaml [hunk 2]
```diff
@@ -14,14 +14,16 @@ spec:
...
+        releaseName: {{ .Release.Name }}
       imagePullSecrets:
         - name: {{ .Release.Name }}-ghcr
       containers:
         - name: {{ .Release.Name }}-helper
           image: {{ .Values.helperImage }}
           imagePullPolicy: Always
           command: ["ipa-helper"]
-          args: ["--identity", "1",
+          args: ["--identity", "{{ trimPrefix "h" .Release.Name }}",
```
Comment (eriktaubeneck) [inline templates/deployment.yaml:26]: why not just having the `releaseName` be the integer?

File: templates/service.yaml [hunk 3]
```diff
@@ -1,11 +1,17 @@
...
 metadata:
-  name: {{ .Release.Name }}-helper-service
+  name: {{ .Release.Name }}-lb
+  annotations:
+    service.beta.kubernetes.io/aws-load-balancer-type: nlb
+    service.beta.kubernetes.io/aws-load-balancer-scheme: internet-facing
+    service.beta.kubernetes.io/aws-load-balancer-nlb-target-type: instance
 spec:
-  type: NodePort
+  type: LoadBalancer
```
Comment (eriktaubeneck) [inline templates/service.yaml:10]: just to clarify, is this a LoadBalancer per helper? thinking a little bit ahead, but I know we have concerns about shards connecting through a load balancer being a choke point (and also, it's not really doing any balancing).

File: templates/deployment.yaml (hunk 2 above)
Comment (cberkhoff) [reply templates/deployment.yaml:26]: I thought about it, but decided against it. "Names" don't usually start with a number and I'm using the releaseName in a bunch of places that expect valid variable names.

File: templates/service.yaml (hunk 3 above)
Comment (cberkhoff) [reply templates/service.yaml:10]: K8s requires some sort of Service to expose the Pods to the internet. From what I investigated, using this kind of LoadBalancer (NLB) suits our needs best.

This is an AWS Elastic Load Balancer, so I don't imagine this being a choke point. Also, this is a "simple" Network Load Balancer. This one only routes on the network layer and can't for instance do SSL termination (that's on a higher level).

File: eksctl/cluster-config.yaml (hunk 1 above)
Comment (cberkhoff) [reply eksctl/cluster-config.yaml:19]: Can I do that on a future change? I'm not quite sure what we need to use.

File: templates/service.yaml (hunk 3 above)
Comment (eriktaubeneck) [reply templates/service.yaml:10]: got it. this is fine for now, but having a single load balancer per helper may become an issue with multiple shards. @akoshelev has brought this up in the past

**Review by eriktaubeneck — APPROVED** [review]

File: templates/service.yaml (hunk 3 above)
Comment (akoshelev) [reply templates/service.yaml:10]: is it really needed? adding extra hops even at L4 does not help with scale. I am a bit surprised that we must have a proxy in front of our hosts but I guess that is K8-specific quirk?

Comment (cberkhoff) [reply templates/service.yaml:10]: This is one of the different options K8s provides. Here are more details: https://kubernetes.io/docs/concepts/services-networking/

I'm of the opinion that having some kind of Load Balancer will be necessary to achieve High Availability and be able to hot switch between hosts if there's a new deployment.

The NodePort approach from before had this "weird" behavior, where a Pod (a helper) was mapped to a specific port (say 30001 for h1, 30002 for h2, etc). Then a call to any of the Nodes (using its public IP) in the cluster would proxy the port to the right host. This meant that I could call h2 from the h1 host if I used port 30002.

Comment (akoshelev) [reply templates/service.yaml:10]: lets sync up between us - there might be an important aspect we're missing when discussing this topic. I want to make sure we are on the same page and understand what we're getting from k8 here

**Pull request merged**